File: biddo_web/src/core/controllers/categories.ts
import { Category } from '../domain/category'
import { AppStore } from '../store'

class CategoriesController {
  getPersonalizedCategoriesForHome = (categories: Category[], categoriesToGet = 5) => {
    if (!categories.length) {
      return []
    }

    if (categories.length <= 8) {
      return categories.slice(0, categoriesToGet)
    }

    const account = AppStore.accountData
    if (!account?.preferredCategoriesIds.length) {
      return this.getRandomCategories(categoriesToGet, categories)
    }

    if (account?.preferredCategoriesIds.length > categoriesToGet) {
      return this.getPreferredCategoriesWithCount(categories, categoriesToGet)
    }

    const neededCount = categoriesToGet - (account?.preferredCategoriesIds?.length ?? 0)
    const newCategories = this.getNextCategories(
      neededCount,
      categories,
      account.preferredCategoriesIds
    )
    return [...newCategories, ...this.getPreferredCategoriesWithCount(categories)]
  }

  private getRandomCategories = (
    n: number,
    categories: Category[],
    fromAllCategories?: boolean
  ) => {
    if (fromAllCategories == true) {
      return this.getRandomFromList(n, categories)
    }

    const categoriesWithAuctions = categories.filter((element) => element.auctionsCount || 0 > 0)

    if (!categoriesWithAuctions.length) {
      return this.getRandomFromList(n, categories)
    }

    if (categoriesWithAuctions.length == n) {
      return categoriesWithAuctions
    }

    if (categoriesWithAuctions.length > n) {
      return this.getRandomFromList(n, categoriesWithAuctions)
    }

    const restOfCategories = this.getRandomFromList(
      n - categoriesWithAuctions.length,
      categories,
      categoriesWithAuctions
    )

    return [...categoriesWithAuctions, ...restOfCategories]
  }

  private getPreferredCategoriesWithCount = (categories: Category[], limit?: number) => {
    if (!AppStore.accountData?.id) {
      return []
    }

    const preferredCategories = AppStore.accountData.preferredCategoriesIds
    const categoriesWithAuctionCount = categories.filter((category) =>
      preferredCategories.some((element) => element == category.id)
    )

    if (!limit) {
      return categoriesWithAuctionCount
    }

    categoriesWithAuctionCount.sort((a, b) => {
      if ((a.auctionsCount ?? 0) > (b.auctionsCount ?? 0)) {
        return -1
      }

      if ((a.auctionsCount ?? 0) < (b.auctionsCount ?? 0)) {
        return 1
      }
      return 0
    })

    return categoriesWithAuctionCount.slice(0, limit)
  }

  private getRandomFromList = (n: number, categoriesList: Category[], except?: Category[]) => {
    const result = [] as Category[]

    for (let i = 0; i < n; i++) {
      const random = Math.floor(Math.random() * categoriesList.length)
      const randomCategory = categoriesList[random]

      const exists = result.some((element) => element.id == randomCategory.id)
      const existsInExcept = except?.some((element) => element.id == randomCategory.id)

      if (!exists && !existsInExcept) {
        result.push(randomCategory)
      } else {
        i -= 1
      }
    }
    return result
  }

  private getNextCategories = (n: number, categories: Category[], except: string[]) => {
    if (n > categories.length - except.length) {
      throw new Error('given param is too big')
    }

    const result = [] as Category[]
    let foundCategoriesCount = 0

    while (foundCategoriesCount < n) {
      const randomCategory = this.getRandomCategories(1, categories, true)[0]
      const existsInExcept = except.some((element) => element == randomCategory.id)
      const existsInResult = result.some((element) => element.id == randomCategory.id)

      if (existsInExcept || existsInResult) {
        continue
      }

      result.push(randomCategory)
      foundCategoriesCount += 1
    }

    return result
  }
}

const categoriesController = new CategoriesController()
export { categoriesController as CategoriesController }

File: biddo_client/lib/core/controllers/categories.dart
import 'dart:math';

import 'package:get/get.dart';

import '../models/category.dart';
import '../repositories/categories.dart';
import 'account.dart';

class CategoriesController extends GetxController {
  final categoriesRepository = Get.find<CategoriesRepository>();
  final accountController = Get.find<AccountController>();

  final RxList<Rx<Category>> categories = <Rx<Category>>[].obs;

  @override
  void onInit() async {
    super.onInit();
    categories.clear();
  }

  @override
  void onClose() {
    super.onClose();
    categories.clear();
  }

  void incrementCategoryAuctionCount(String categoryId) {
    var category = categories.firstWhereOrNull(
      (element) => element.value.id == categoryId,
    );
    if (category == null) {
      return;
    }
    category.value.auctionsCount += 1;
    category.refresh();
  }

  void decrementCategoryAuctionCount(String categoryId) {
    var category = categories.firstWhereOrNull(
      (element) => element.value.id == categoryId,
    );
    if (category == null) {
      return;
    }
    category.value.auctionsCount -= 1;
    category.refresh();
  }

  Future<void> load() async {
    var categoriesList = await categoriesRepository.loadAll();
    categories.clear();
    categories.addAll(categoriesList.map((category) => category));
  }

  int getNumberOfCardsForHome() {
    return min(6, categories.length);
  }

  List<Rx<Category>> getPersonalizedCategoriesForHome() {
    if (categories.isEmpty) {
      return [];
    }

    var numberOfCards = getNumberOfCardsForHome();
    var account = accountController.account.value;
    if (account.preferredCategoriesIds.isEmpty) {
      return getRandomCategories(numberOfCards);
    }

    if (account.preferredCategoriesIds.length > numberOfCards) {
      return getPreferredCategoriesWithCount(numberOfCards);
    }

    var neededCount = numberOfCards - account.preferredCategoriesIds.length;
    var newCategories =
        getNextCategories(neededCount, account.preferredCategoriesIds);
    return [...newCategories, ...getPreferredCategoriesWithCount()];
  }

  Category? findById(String id) {
    var category =
        categories.firstWhereOrNull((category) => category.value.id == id);

    if (category == null) {
      // search in subcategories
      return categories
          .map((category) => category.value.subcategories)
          .expand((element) => element)
          .firstWhere((subcategory) => subcategory.value.id == id)
          .value;
    }

    return category.value;
  }

  List<Rx<Category>> getRandomCategories(int n, [bool? fromAllCategories]) {
    if (fromAllCategories == true) {
      return getRandomFromList(n, categories);
    }

    var categoriesWithAuctions =
        categories.where((element) => element.value.auctionsCount > 0).toList();

    if (categoriesWithAuctions.isEmpty) {
      return getRandomFromList(n, categories);
    }

    if (categoriesWithAuctions.length == n) {
      return categoriesWithAuctions;
    }

    if (categoriesWithAuctions.length > n) {
      return getRandomFromList(n, categoriesWithAuctions);
    }

    var restOfCategories = getRandomFromList(
      n - categoriesWithAuctions.length,
      categories,
      categoriesWithAuctions,
    );

    return [...categoriesWithAuctions, ...restOfCategories];
  }

  List<Rx<Category>> getRandomFromList(
    int n,
    List<Rx<Category>> categoriesList, [
    List<Rx<Category>> except = const [],
  ]) {
    List<Rx<Category>> result = [];

    for (var i = 0; i < n; i++) {
      Random rand = Random();
      var randomCategory = categoriesList[rand.nextInt(categoriesList.length)];

      var exists = result
          .where((element) => element.value.id == randomCategory.value.id);
      var existsInExcept = except
          .where((element) => element.value.id == randomCategory.value.id);

      if (exists.isEmpty && existsInExcept.isEmpty) {
        result.add(randomCategory);
      } else {
        i -= 1;
      }
    }
    return result;
  }

  List<Rx<Category>> getPreferredCategoriesWithCount([int? limit]) {
    var preferredCategories =
        accountController.account.value.preferredCategoriesIds;
    var categoriesWithAuctionCount = categories
        .where(
          (category) => preferredCategories.any(
            (element) => element == category.value.id,
          ),
        )
        .toList();

    if (limit == null) {
      return categoriesWithAuctionCount;
    }

    categoriesWithAuctionCount.sort((a, b) {
      if (a.value.auctionsCount > b.value.auctionsCount) {
        return -1;
      }

      if (a.value.auctionsCount < b.value.auctionsCount) {
        return 1;
      }
      return 0;
    });

    return categoriesWithAuctionCount.take(limit).toList();
  }

  List<Rx<Category>> getNextCategories(int n, List<String> except) {
    if (n > (categories.length - except.length)) {
      throw Exception('given param is too big');
    }

    List<Rx<Category>> result = [];
    var foundCategoriesCount = 0;

    while (foundCategoriesCount < n) {
      var randomCategory = getRandomCategories(1, true)[0];

      var existsInExcept =
          except.any((element) => element == randomCategory.value.id);
      var existsInResult =
          result.any((element) => element.value.id == randomCategory.value.id);
      if (existsInExcept || existsInResult) {
        continue;
      }

      result.add(randomCategory);
      foundCategoriesCount += 1;
    }

    return result;
  }
}
